server/server.go

```diff
@@ -0,0 +1,86 @@
+package server
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/alphagov/paas-usage-events-collector/api"
+	"github.com/alphagov/paas-usage-events-collector/auth"
+	"github.com/alphagov/paas-usage-events-collector/db"
+	"github.com/labstack/echo"
+	"github.com/labstack/echo/middleware"
+	"github.com/labstack/gommon/log"
+)
+
+func New(db db.SQLClient, authority auth.Authenticator) *echo.Echo {
+	e := echo.New()
+	e.Logger.SetLevel(log.INFO)
+
+	// Allow HTML forms to override POST method using _method param
+	e.Pre(middleware.MethodOverrideWithConfig(middleware.MethodOverrideConfig{
+		Getter: middleware.MethodFromForm("_method"),
+	}))
+
+	// Never crash on panic
+	e.Use(middleware.Recover())
+
+	// Require a token for all requests
+	e.Use(auth.UAATokenAuthentication(authority))
+
+	// Validate and parse range query param
+	e.Use(api.ValidateRangeParams)
+
+	// Deprecated endpoint, favor /resources and /events
+	e.GET("/usage", api.NewUsageHandler(db))
+
+	// An example billing renderer... throw this away
+	e.GET("/report", api.NewReportHandler(db))
+	e.GET("/", redirectToReport)
+
+	// Usage and Billing API
+	e.GET("/organisations", api.ListOrgUsage(db))
+	e.GET("/organisations/:org_guid", api.GetOrgUsage(db))
+	e.GET("/organisations/:org_guid/spaces", api.ListSpacesUsageForOrg(db))
+	e.GET("/organisations/:org_guid/resources", api.ListResourceUsageForOrg(db))
+	e.GET("/spaces", api.ListSpacesUsage(db))
+	e.GET("/spaces/:space_guid", api.GetSpaceUsage(db))
+	e.GET("/spaces/:space_guid/resources", api.ListResourceUsageForSpace(db))
+	e.GET("/resources", api.ListResourceUsage(db))
+	e.GET("/resources/:resource_guid", api.GetResourceUsage(db))
+	e.GET("/events", api.ListEventUsage(db))
+	e.GET("/resources/:resource_guid/events", api.ListEventUsageForResource(db))
+
+	// Pricing data API
+	e.GET("/pricing_plans", api.ListPricingPlans(db))
+	e.GET("/pricing_plans/:pricing_plan_id", api.GetPricingPlan(db))
+	e.POST("/pricing_plans", auth.AdminOnly(api.CreatePricingPlan(db)))
+	e.PUT("/pricing_plans/:pricing_plan_id", auth.AdminOnly(api.UpdatePricingPlan(db)))
+	e.DELETE("/pricing_plans/:pricing_plan_id", auth.AdminOnly(api.DestroyPricingPlan(db)))
+	e.POST("/seed_pricing_plans", auth.AdminOnly(api.CreateMissingPricingPlans(db)))
+
+	return e
+}
+
+func redirectToReport(c echo.Context) error {
+	lastMonth := time.Now().UTC().Add(-30 * 24 * time.Hour).Format(time.RFC3339)
```

Comment:
This could be a bit more user friendly by selecting the first day of the month:

```
now := time.Now().UTC()
lastMonth := time.Date(now.Year(), now.Month()-1, 1, 0, 0, 0, 0, now.Location()).Format(time.RFC3339)
```

Reply:
Unless a rolling 30-day thing was what you were after

Reply:
I was going for a "show me recent usage to get started" type affair, so there was some example date ranges to play with.

The whole "report" and html ui thing is a bit of a sketch tbh ...it's more of a "you could make a UI that does things like this" than a "this is a ui we are going to keep". It was added because it was tiresome having to make API calls directly to play with the output and because without it I didn't really see a way for Tom to poke around.

I suspect all this functionality to end up in the propper "paas admin dashboard/control-panel" ruby app thing eventully.

Reply:
Cool. Leave as is. We can do code tweaks when people start actually using the thing.